### PR TITLE
MGMT-1176: Onprem: Configure agent with dns so it can contact assisted service

### PR DESCRIPTION
**Pull request by sacharya**
In an onprem environment, agent is configured with a list of service ips
and hostname in its etc/hosts file, so it can talk to the service endpoint.

### CODE REVIEW

**Comment by sacharya** [comment]
Had to create a new PR based on Pawan's https://github.com/openshift/assisted-service/pull/182

File: config/onprem-iso-config.ign [hunk 1]
```diff
@@ -21,15 +21,15 @@
         "path": "/etc/assisted-service/create_certs.sh",
         "contents": {
           "compression": "gzip",
-          "source": "data:;base64,H4sIAAAAAAAC/5zPsW6DMBAG4J2nuNL5cKCAklQMiClL2qp9AWNfExdjUt+lhbevImXqEinbDfd/v/7HB9W7oHrNx2QcrIuAJ1AkRmlmx0IWmeKPM6TCwYUZDUXhZDpRYPYQ6RtwrlYbwDBZYkCrF4anugLkc/8FqeqarlXvH81bp16a9orCLrBo7ymqbt/kmyLL63WWF0VWpYDaWpoF0otARlovez1Ss3vd/v8M9DvQApH1tliVa8CBlukstxdcbyb/ye4QyGYXB+/MmijPiTmOk4W6LO9sT/4CAAD//9U0UEGMAQAA"
+          "source": "data:;base64,H4sIAAAAAAAC/5yPsU7DMBCG9zzFEeaLK0grKMoQhYUlCJUXcOxrY+LYwXeF5O1RUdW1Urd/+b5P//2d6lxQneY+GwfrEuAEisQozexYyCJT+nGGVDi4MKOhJJzFiQKzh0TfgPN69QwYoiUGtHpheNysAfnYfUGumqqp1e6z+mjUe1WfpfAWWLT3lFTTVpeUnlzho9G++A/0bi+FizmgtpZmgfzkJCO1l1aPVL22u+11ONDvQAsk1tuHVfkEONASj3L95nkz+T27QyBbnDx4I2uSvGSmH6OFTVneWM/+AgAA//9PXnJesQEAAA=="
```
Comment (carbonin) [inline config/onprem-iso-config.ign:24]: All of these scripts should be able to be reviewed as well. Is there some better way to store these so that I don't have to unpack all of this every time new changes come in?

Maybe we should post the contents or the patch in a comment if we really have to keep these in this form?
@mhrivnak any ideas?

Comment (sacharya) [reply config/onprem-iso-config.ign:24]: The .ign file is actually generated from fcc.yaml, which has the readable content.

File: config/onprem-iso-config.ign [hunk 2]
```diff
@@ -58,7 +66,12 @@
         "name": "assisted-service-certs.service"
       },
       {
-        "contents": "[Unit]\nAfter=network-online.target\n\n[Service]\nType=forking\nRestart=no\nExecStart=podman pod create --name assisted-installer -p 8443\n\n[Install]\nWantedBy=multi-user.target\n",
+        "contents": "[Unit]\nAfter=network-online.target\n\n[Service]\nType=forking\nRestart=no\nExecStart=/bin/bash /etc/assisted-service/startup_script.sh\n\n[Install]\nWantedBy=multi-user.target\n",
+        "enabled": true,
+        "name": "assisted-service-ip-configuration.service"
+      },
+      {
+        "contents": "[Unit]\nAfter=assisted-service-ip-configuration.service\n\n[Service]\nType=forking\nRestart=no\nExecStart=podman pod create --name assisted-installer -p 8443\n\n[Install]\nWantedBy=multi-user.target\n",
```
Comment (carbonin) [inline config/onprem-iso-config.ign:74]: I think this actually creates a race condition. 
`After=` only actually guarantees that this unit will start after the specified one is started. There is no guarantee that we will finish writing out the environment file in `assisted-service-ip-configuration` before this unit is started.

That said it's hard to follow the flow here as I have to decode all of the script content. This is what lead to my other comment :smile:

File: config/onprem-iso-config.ign (hunk 1 above)
Comment (carbonin) [reply config/onprem-iso-config.ign:24]: Lol just realised this is generated and the scripts are below. :facepalm:

File: config/onprem-iso-config.ign (hunk 2 above)
Comment (carbonin) [reply config/onprem-iso-config.ign:74]: Nope, was confusing `After` with `Wants` ... :tired_face:

File: config/onprem-iso-fcc.yaml [hunk 3]
```diff
@@ -129,3 +142,16 @@ storage:
...
+            inline: |
+              #!/bin/bash
+              ips=$(hostname -I)
+              read -r -a ipArr <<< "$ips"
+              for ip in "${ipArr[@]}"
+                  do 
+                      printf "\n%s assisted-api.local.openshift.io\n" "$ip" >> /etc/hosts
+                  done
+              ips_cs=`echo $ips | xargs | sed -e 's/ /,/g'`
+              printf "\nSERVICE_IPS=%s" "$ips_cs" >> /etc/assisted-service/environment
```
Comment (carbonin) [inline config/onprem-iso-fcc.yaml:157]: I wish there was a way to test this :worried: 

Also, if I'm reading this correctly, we are going to set the same hostname to every available ip on the host...
That seems strange, is that really what we want to do?

Comment (sacharya) [reply config/onprem-iso-fcc.yaml:157]: Yeah, it's a bit of a hack :exploding_head: We needed a way to get the IP of host running the service that is routable from the agent nodes, without making it yet another thing user has to provide. This will configure the discovery iso ignition with all the ips of the service host and the agent will try all the ips until it finds one that is routable form node where the agent iso is being booted.

Comment (carbonin) [reply config/onprem-iso-fcc.yaml:157]: > the agent will try all the ips until it finds one that is routable form node where the agent iso is being booted

Where is this logic?

Comment (carbonin) [reply config/onprem-iso-fcc.yaml:157]: I'm mostly concerned that this is being added to `/etc/hosts` and as far as I can tell we will only ever resolve the hostname to the first matching entry in the file. If we need this special logic I don't think we can use `/etc/hosts`.

Comment (carbonin) [reply config/onprem-iso-fcc.yaml:157]: Nice, thanks @mhrivnak TIL.

Still kinda messy but makes sense now. :+1:

**Comment by carbonin** [comment]
@sacharya what is the relationship between this PR and https://github.com/openshift/assisted-installer/pull/61?
If this gets merged first, will installs fail because we are providing an option that doesn't exist yet? Should this be WIP until https://github.com/openshift/assisted-installer/pull/61 is merged?

**Comment by sacharya** [comment]
@carbonin I think they should be reviewed in parallel as the solution would be incomplete without one of them. But, technically https://github.com/openshift/assisted-installer/pull/61 should be merged before this one.  I can hold or WIP this one, but wanted to get rough approval on this approach before getting https://github.com/openshift/assisted-installer/pull/61  merged.

**Comment by carbonin** [comment]
/hold

Putting this on hold until https://github.com/openshift/assisted-installer/pull/61 is merged

**Review by carbonin — APPROVED** [review]
